packages/jazz-run/src/startSync.ts

```diff
@@ -0,0 +1,121 @@
+        return Effect.gen(function* () {
+            const crypto = yield* Effect.promise(() => WasmCrypto.create());
+
+            const wss = new WebSocketServer({ port: parseInt(port) });
+
+            console.log(
+                "COJSON sync server listening on port " + wss.options.port,
+            );
+
+            /**
+             * In Jazz the communication is decentralized
+             * so the sync server is implemented as "just another peer" in the network
+             *
+             * LocalNode is the class we use to manage the connected peers that could be:
+             * - the storage layer
+             * - the WebSocket connections
+             */
+            const agentSecret = crypto.newRandomAgentSecret();
+            const agentID = crypto.getAgentID(agentSecret);
+
+            const localNode = new LocalNode(
+                new ControlledAgent(agentSecret, crypto),
+                cojsonInternals.newRandomSessionID(agentID),
+                crypto,
+            );
+
+            if (!inMemory) {
+                yield* Effect.promise(() =>
+                    mkdir(dirname(db), { recursive: true }),
+                );
+
+                const storage = yield* Effect.promise(() =>
+                    SQLiteStorage.asPeer({ filename: db }),
+                );
+
+                localNode.syncManager.addPeer(storage);
+            }
+
+            wss.on("connection", function connection(ws, req) {
+                // ping/pong for the connection livenerss
+                const pinging = setInterval(() => {
+                    ws.send(
+                        JSON.stringify({
+                            type: "ping",
+                            time: Date.now(),
+                            dc: "unknown",
+                        }),
+                    );
+                }, 1500);
+
+                ws.on("close", () => {
+                    clearInterval(pinging);
+                });
+
+                const clientAddress =
+                    (req.headers["x-forwarded-for"] as string | undefined)
+                        ?.split(",")[0]
+                        ?.trim() || req.socket.remoteAddress;
+
+                const clientId = clientAddress + "@" + new Date().toISOString();
+
+                localNode.syncManager.addPeer(
+                    createWebSocketPeer({
+                        id: clientId,
+                        role: "client",
+                        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+                        websocket: ws as any, // TODO: fix types
+                        expectPings: false,
+                    }),
+                );
+
+                ws.on("error", (e) =>
+                    console.error(`Error on connection ${clientId}:`, e),
+                );
+            });
+
+            // Indefintely pending promise to keep the server up
+            yield* Effect.promise(() => new Promise(() => {}));
```

Comment:
Asking some of the Effect pros I know what to best do here

Reply:
can also do `yield* Effect.never` I guess

Reply:
I'd probably use Layer.launch and launch a SocketServer here (see @effect/experimental SocketServer.layerWebSocket). So basically creating the websocket server through Effect instead of via ws manually.

You'd likely then also want to utilize Effect for the file system interaction (creating the database, etc.) here.

That said, I'm not sure if that's in scope for this PR and if that's leading you on a bunch of tangents and Effect exploration journeys that you might not be ready for / willing to go on for the sake of this PR ;-)

Reply:
Thanks for the suggestions!

Gone with `yield* Effect.never` as doing more isn't within the scope of this PR (the goal was to land with a minimal implementation of a sync server)

Interested on learning how to use Effect for the file system interactions / websocket server.

Are there any examples or docs to look at to get into it? @fubhy

Reply:
Thanks @Schniz for the quick suggestion

Thanks @fubhy for the more extensive roadmap - out of scope for this PR but exactly what I had in mind when we have the time to re-adopt Effect more meaningfully

Reply:
> Thanks for the suggestions!
> 
> Gone with `yield* Effect.never` as doing more isn't within the scope of this PR (the goal was to land with a minimal implementation of a sync server)
> 
> Interested on learning how to use Effect for the file system interactions / websocket server.
> 
> Are there any examples or docs to look at to get into it? @fubhy

That makes sense.

There are no examples for the SocketServer afaik. But there's an implementation of a SocketServer in the DevTools package within the Effect monorepo.

Either way, I'd suggest to hit our Discord server once you start diving more into Effect. That's the best place to ask questions and learn.

Regarding the FileSystem module there are some tests here that might give you an idea: https://github.com/Effect-TS/effect/blob/main/packages/platform-node-shared/test/FileSystem.test.ts